File: internal/documents/content_blocks.go
package documents

import (
	"encoding/base64"
	"strings"

	"github.com/trknhr/slack-ai-assistant/internal/anthropic"
)

func BuildClaudeContentBlocksForDocument(title string, mimeType string, content []byte) []anthropic.InputBlock {
	if mimeType == "application/pdf" {
		return []anthropic.InputBlock{
			{
				"type":  "document",
				"title": title,
				"source": map[string]any{
					"type":       "base64",
					"media_type": "application/pdf",
					"data":       base64.StdEncoding.EncodeToString(content),
				},
			},
		}
	}
	if len(mimeType) >= 6 && mimeType[:6] == "image/" {
		return []anthropic.InputBlock{
			{
				"type": "text",
				"text": "Attached image: " + title,
			},
			{
				"type": "image",
				"source": map[string]any{
					"type":       "base64",
					"media_type": mimeType,
					"data":       base64.StdEncoding.EncodeToString(content),
				},
			},
		}
	}
	if isTextLikeMimeType(mimeType) {
		return []anthropic.InputBlock{
			{
				"type":  "document",
				"title": title,
				"source": map[string]any{
					"type":       "text",
					"media_type": "text/plain",
					"data":       string(content),
				},
			},
		}
	}
	return []anthropic.InputBlock{
		{
			"type": "text",
			"text": "Attachment note: " + title + " (" + chooseMimeType(mimeType) + ") is not supported for inline analysis.",
		},
	}
}

func chooseMimeType(value string) string {
	if value == "" {
		return "unknown mime type"
	}
	return value
}

func isTextLikeMimeType(mimeType string) bool {
	return strings.HasPrefix(mimeType, "text/") ||
		mimeType == "application/json" ||
		mimeType == "application/xml" ||
		mimeType == "application/javascript"
}

File: src/documents/contentBlocks.ts
import { AgentContentBlock } from "../agent/types";
import { isTextLikeMimeType } from "../slack/fileSupport";

export function buildAgentContentBlocksForDocument(
  title: string,
  mimeType: string | undefined,
  buffer: Buffer,
): AgentContentBlock[] {
  if (mimeType === "application/pdf") {
    return [
      {
        type: "document",
        title,
        source: {
          type: "base64",
          media_type: "application/pdf",
          data: buffer.toString("base64"),
        },
      },
    ];
  }

  if (mimeType && mimeType.startsWith("image/")) {
    return [
      {
        type: "text",
        text: `Attached image: ${title}`,
      },
      {
        type: "image",
        source: {
          type: "base64",
          media_type: mimeType,
          data: buffer.toString("base64"),
        },
      },
    ];
  }

  if (isTextLikeMimeType(mimeType)) {
    return [
      {
        type: "document",
        title,
        source: {
          type: "text",
          media_type: "text/plain",
          data: buffer.toString("utf-8"),
        },
      },
    ];
  }

  return [
    {
      type: "text",
      text: `Attachment note: ${title} (${mimeType ?? "unknown mime type"}) is not supported for inline analysis.`,
    },
  ];
}
